### PR TITLE
Migrate from `anyhow` to `thiserror`

### PR DESCRIPTION
**Pull request by mhovd**
This PR migrates `diffsol` from `anyhow` to `thiserror`. This allows for matchable errors for third-party users.

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
+use faer::sparse::CreationError;
+use thiserror::Error;
+
+/// Custom error type for Diffsol
+///
+/// This error type is used to wrap all possible errors that can occur when using Diffsol
+#[derive(thiserror::Error, Debug)]
```
Comment (martinjrobins) [inline src/error.rs:7]:
```suggestion
#[derive(Error, Debug)]
```

File: src/error.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+pub enum DiffsolError {
+    #[error("Linear solver error: {0}")]
+    LinearSolverError(#[from] LinearSolverError),
+    #[error("Non-linear solver error: {0}")]
+    NonLinearSolverError(#[from] NonLinearSolverError),
+    #[error("ODE solver error: {0}")]
+    OdeSolverError(#[from] OdeSolverError),
+    #[error("Matrix error: {0}")]
+    MatrixError(#[from] MatrixError),
+    #[error("Other error: {0}")]
```
Comment (martinjrobins) [inline src/error.rs:17]:
```suggestion
    #[error("Error: {0}")]
```

File: src/error.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    LuNotInitialized,
+    #[error("LU solve failed")]
+    LuSolveFailed,
+    #[error("Linear solver not setup")]
+    LinearSolverNotSetup,
+    #[error("KLU failed to analyze")]
+    KluFailedToAnalyze,
+    #[error("KLU failed to factorize")]
+    KluFailedToFactorize,
+    #[error("Other error: {0}")]
```
Comment (martinjrobins) [inline src/error.rs:34]:
```suggestion
    #[error("Error: {0}")]
```

File: src/error.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    Other(String),
+}
+
+/// Possible errors that can occur when solving a non-linear problem
+#[derive(Error, Debug)]
+pub enum NonLinearSolverError {
+    #[error("Newton did not converge")]
```
Comment (martinjrobins) [inline src/error.rs:41]:
```suggestion
    #[error("Newton iterations did not converge")]
```

File: src/error.rs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    NewtonDidNotConverge,
+    #[error("LU solve failed")]
+    LuSolveFailed,
+    #[error("Other error: {0}")]
```
Comment (martinjrobins) [inline src/error.rs:45]:
```suggestion
    #[error("Error: {0}")]
```

File: src/error.rs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    FailedToGetMutableReference,
+    #[error("atol must have length 1 or equal to the number of states")]
+    AtolLengthMismatch,
+    #[error("t_eval must be increasing and all values must be greater than or equal to the current time")]
+    InvalidTEval,
+    #[error("Sundials error: {0}")]
+    SundialsError(String),
+    #[error("Problem not set")]
+    ProblemNotSet,
+    #[error("Other error: {0}")]
```
Comment (martinjrobins) [inline src/error.rs:82]:
```suggestion
    #[error("Error: {0}")]
```

File: src/error.rs [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+pub enum MatrixError {
+    #[error("Failed to create matrix from triplets: {0}")]
+    FailedToCreateMatrixFromTriplets(#[from] CreationError),
+    #[error("Cannot union matrices with different shapes")]
+    UnionIncompatibleShapes,
+    #[error("Cannot create a matrix with zero rows or columns")]
+    MatrixShapeError,
+    #[error("Index out of bounds")]
+    IndexOutOfBounds,
+    #[error("Other error: {0}")]
```
Comment (martinjrobins) [inline src/error.rs:97]:
```suggestion
    #[error("Error: {0}")]
```

File: src/ode_solver/sundials.rs [hunk 8]
```diff
@@ -407,30 +408,30 @@ where
...
-            IDA_CONV_FAIL => Err(anyhow!("Convergence test failures occurred too many times (MXNCF = 10) during one internal time step or occurred with.")),
-            IDA_LINIT_FAIL => Err(anyhow!("The linear solver’s initialization function failed.")),
-            IDA_LSETUP_FAIL => Err(anyhow!("The linear solver’s setup function failed in an unrecoverable manner.")),
-            IDA_LSOLVE_FAIL => Err(anyhow!("The linear solver’s solve function failed in an unrecoverable manner.")),
-            IDA_CONSTR_FAIL => Err(anyhow!("The inequality constraints were violated and the solver was unable to recover.")),
-            IDA_REP_RES_ERR => Err(anyhow!("The user’s residual function repeatedly returned a recoverable error flag, but the solver was unable to recover.")),
-            IDA_RES_FAIL => Err(anyhow!("The user’s residual function returned a nonrecoverable error flag.")),
-            IDA_RTFUNC_FAIL => Err(anyhow!("The rootfinding function failed.")),
-            _ => Err(anyhow!("Unknown error")),
+            IDA_MEM_NULL => Err(other_error!("The ida_mem argument was NULL.")),
```
Comment (martinjrobins) [inline src/ode_solver/sundials.rs:411]: should these all be `OdeSolverError::SundialsError` as well?

**Review by martinjrobins — CHANGES_REQUESTED** [review]
Thanks @mhovd, this looks great. Some suggestions below (mainly minor and pedantic)

File: src/ode_solver/sundials.rs (hunk 8 above)
Comment (mhovd) [reply src/ode_solver/sundials.rs:411]: Sure, I just didn't know if the those from `IDA` was regarded as such.

Comment (martinjrobins) [reply src/ode_solver/sundials.rs:411]: Yup, IDA is a particular solver within the Sundials solver suite

Comment (mhovd) [reply src/ode_solver/sundials.rs]: Fixed

**Review by martinjrobins — APPROVED** [review]
thanks @mhovd, this looks good, just some error mapping required for the diffsl errors then this can be merged

**Pull request merged**